tools/java/runfiles/Runfiles.java

```diff
@@ -183,20 +340,58 @@ private static String getRunfilesDir(Map<String, String> env) throws IOException
     return value;
   }
 
+  private static Map<RepoMappingKey, String> loadRepositoryMapping(String path)
+      throws IOException {
+    if (path == null) {
+      return Collections.emptyMap();
+    }
+
+    try (BufferedReader r = new BufferedReader(new FileReader(path, StandardCharsets.UTF_8))) {
+      return Collections.unmodifiableMap(r.lines()
+          .filter(line -> !line.isEmpty())
+          .map(line -> {
+            String[] split = line.split(",");
+            if (split.length != 3) {
+              throw new IllegalArgumentException(
+                  "Invalid line in repository mapping: '" + line + "'");
+            }
+            return split;
+          })
+          .collect(Collectors.toMap(
+              split -> new RepoMappingKey(split[0], split[1]),
+              split -> split[2])));
+    }
+  }
+
   abstract String rlocationChecked(String path);
 
-  /** {@link Runfiles} implementation that parses a runfiles-manifest file to look up runfiles. */
+  /**
+   * {@link Runfiles} implementation that parses a runfiles-manifest file to look up runfiles.
+   */
   private static final class ManifestBased extends Runfiles {
+
     private final Map<String, String> runfiles;
     private final String manifestPath;
 
-    ManifestBased(String manifestPath) throws IOException {
+    ManifestBased(String manifestPath, String sourceRepository) throws IOException {
+      super(findRepositoryMapping(manifestPath), sourceRepository);
       Util.checkArgument(manifestPath != null);
       Util.checkArgument(!manifestPath.isEmpty());
       this.manifestPath = manifestPath;
       this.runfiles = loadRunfiles(manifestPath);
     }
 
+    private ManifestBased(ManifestBased existing, String sourceRepository) {
+      super(existing, sourceRepository);
+      this.manifestPath = existing.manifestPath;
+      this.runfiles = existing.runfiles;
+    }
+
+    @Override
+    public Runfiles withSourceRepository(String sourceRepository) {
```

Comment:
What's the use case? Please remove.

This is an optimisation, that I doubt anybody will have a need to use. As such it just adds maintenance burden.

Reply:
This is meant to preserve the common pattern of injecting a `Runfiles` instance at the top and feeding it to downstream consumers with Bzlmod. By using this function, users only have to make a change at one central location per repository (where they need to pass in their repository name obtained via the annotation), rather than at every `Rlocation` invocation outside the main repository.

In particular, this makes it possible to fully adopt Bzlmod *without* all of your transitive runfiles-using dependencies having done so. I thus consider it a relatively important part of the migration story.

Do you see a way I could address some of your concerns while keeping this method around?

Reply:
Ack. Let's keep conversation about this in a single thread. Resolving this one.